syntax = "proto3";
package Protocol;

option csharp_namespace = "Google.Protobuf.Protocol";

enum MsgId 
{
	C_JWT_LOGIN_REQUEST = 0;
	S_JWT_LOGIN_REPLY = 1;
	C_CREATE_CHARACTER_REQUEST = 2;
	S_CREATE_CHARACTER_REPLY = 3;
	C_CHARACTER_LIST_REQUEST = 4;
	S_CHARACTER_LIST_REPLY = 5;
	C_ENTER_GAME = 6;
	S_ENTER_GAME = 7;
	S_PLAYER_LIST = 8;
	S_BROADCAST_PLAYER_ENTER = 9;
	C_LEAVE_GAME = 10;
	S_LEAVE_GAME = 11;
	S_BROADCAST_PLAYER_LEAVE = 12;
	C_PLAYER_MOVE_REQUEST = 13;
	S_PLAYER_MOVE_REPLY = 14;
	S_BROADCAST_PLAYER_MOVE = 15;
	S_CHANGE_ROOM_BEGIN = 16;
	C_CHANGE_ROOM_READY = 17;
	S_CHANGE_ROOM_COMMIT = 18;
	S_SPAWN_MONSTER = 19;
	S_DESPAWN_MONSTER = 20;
	S_BROADCAST_MONSTER_MOVE = 21;
	S_BROADCAST_MONSTER_ATTACK = 22;
	S_BROADCAST_MONSTER_DEATH = 23;
}

// 클라 -> 게임 서버 jwt 로그인 인증 (0)
message C_JwtLoginRequest 
{
	string accessToken = 1; 
}

// 게임서버 -> 클라 jwt 검증 (1)
message S_JwtLoginReply
{
	ELoginResult result = 1;
}

// (2)
message C_CreateCharacterRequest
{
	string username = 1; // 캐릭터 이름 한글만 가능 2~6자
	EGender gender = 2; 
	ERegion region = 3;
}

// (3)
message S_CreateCharacterReply
{
	bool success = 1;
	string detail = 2; // 실패 시 에러 메시지
}

// (4)
message C_CharacterListRequest
{

}

// (5)
message S_CharacterListReply
{
	repeated CharacterSummaryInfo characters = 1;
}

// (6)
message C_EnterGame
{
	int32 playerIndex = 1;
}

// (7)
message S_EnterGame
{
	int32 success = 1;
}

// (8)
message S_PlayerList
{
	int32 myPlayerId = 1;
	repeated PlayerInfo players = 2;
}

// (9)
message S_BroadcastPlayerEnter
{
	PlayerInfo player = 1;
}

// (10)
message C_LeaveGame
{
	ELeaveReason reason = 1;
}

// (11)
message S_LeaveGame
{
	int32 success = 1;
}

// (12)
message S_BroadcastPlayerLeave
{
	int32 playerId = 1;
	ELeaveReason reason = 2;
}
// (13)
message C_PlayerMoveRequest
{
	Vector2Info clickWorldPos = 1;
}

// (14)
message S_PlayerMoveReply
{
	int32 playerId = 1;
	int32 clientSeq = 2;
	EMoveResult result = 3;
	int32 tick = 4;
	Vector2Info newPos = 5;
	EDirection direction = 6;
}

// (15)
message S_BroadcastPlayerMove
{
	int32 tick = 1;
	repeated PlayerMoveInfo playerMoves = 2; // 변경된 플레이어만 담을 것
}

// (16) 서버가 시작 알림 (클라 입력 잠금, 로딩 UI)
message S_ChangeRoomBegin {
  int32 transitionId = 1;
  int32 mapId        = 2;
}

// (17) 클라가 준비 완료 통보(리소스/타일맵 로드 완료)
message C_ChangeRoomReady {
  int32 transitionId = 1;
}

// (18) 서버가 최종 스폰 좌표/스냅샷 확정
message S_ChangeRoomCommit {
  int32 transitionId = 1;
  S_PlayerList snapshots = 2; // 들어오자마자 월드 동기화
}

// (19) 서버 -> 클라 : 몬스터 스폰 알림
message S_SpawnMonster {
	int32 monsterId = 1; // Entity Unique id
	int32 monsterTypeId = 2; // Monster Type (Template or Resources Id)
	int32 x = 3; // Tile Pos X
	int32 y = 4; // Tile Pos Y
	EDirection dir = 5; // 방향 
}

// (20) 서버 -> 클라 몬스터 디스폰(제거) 알림 
message S_DespawnMonster {
	int32 monsterId = 1;
	EDespawnReason reason = 2; // 정수 코드 EDespawnReason 참고
}

// (21) 서버 -> 클라 : 몬스터 이동 브로드 캐스트
message S_BroadcastMonsterMove {
	int32 monsterId = 1;
	int32 x			= 2; // 이동 위치 X
	int32 y			= 3; // 이동 위치 Y
	EDirection dir	= 4; // 방향
}

// (22) 서버 -> 클라 : 몬스터 공격 브로드 캐스트
message S_BroadcastMonsterAttack {
	int32 monsterId = 1;
	int32 targetPid = 2; // 공격 대상 Player Id
}

// (23) 서버 -> 클라 : 몬스터 사망 브로드 캐스트
message S_BroadcastMonsterDeath {
	int32 monsterId = 1;
}


enum ELoginResult
{
    SUCCESS = 0;
    INVALID_TOKEN = 1;
    TOKEN_EXPIRED = 2;
    SERVER_ERROR = 3;
}

enum EGender 
{
	GENDER_NONE = 0;
	GENDER_MALE = 1;
	GENDER_FEMALE = 2;
}

enum ERegion 
{
    REGION_NONE = 0;
    REGION_GO = 1;
    REGION_BACK = 2;
}

enum EDirection
{
	DIR_UP = 0;
	DIR_DOWN = 1;
	DIR_LEFT = 2;
	DIR_RIGHT = 3;
}

enum ELeaveReason
{
	LEAVE_UNKNOWN = 0;
	LEAVE_LOGOUT = 1; // 게임종료
	LEAVE_CHANGE_ROOM = 2; // 룸 이동
	LEAVE_DISCONNECT = 3; // 네트워크 끊김
}

enum EMoveResult
{
	OK = 0;
	COOLDOWN = 1;
	BLOCKED = 2;
}

enum EEnterReason
{
	ENTER_UNKNOWN = 0;
	ENTER_LOGIN = 1;
	ENTER_CHANGE_ROOM = 2;
}

enum EDespawnReason
{
	UNKNOWN = 0;
	KILLED = 1;
}
message Vector2Info
{
	int32 x = 1;
	int32 y = 2;
}

message PlayerMoveInfo
{
	int32 playerId = 1;
	EDirection direction = 2; 
	Vector2Info newPos = 3;
}

message CharacterSummaryInfo
{
	string username = 1;
	int32 level = 2;
	EGender gender = 3;
	ERegion region = 4;
}

message PlayerInfo
{
	int32 id = 1;
	string username = 2;

	Vector2Info pos = 3;
	EDirection direction = 4;
}